README.md
# story2sketch 💎

[![NPM](https://img.shields.io/npm/v/story2sketch.svg)](https://www.npmjs.com/package/story2sketch) [![JavaScript Style Guide](https://img.shields.io/badge/code_style-prettier-brightgreen.svg)](https://prettier.io)

Convert [Storybook](https://storybook.js.org) stories into [Sketch](https://www.sketchapp.com) symbols.

> Uses the amazing [`html-sketchapp`](https://github.com/html-sketchapp/html-sketchapp). Only supports web.

## Quickstart

Firstly, get [Sketch](https://sketchapp.com) and [npm](https://nodejs.org/en/download/). Then install [`asketch2sketch.sketchplugin`](https://github.com/html-sketchapp/html-sketchapp/releases/download/v4.4.1/asketch2sketch-4-4-1.sketchplugin.zip) into Sketch:

<img src="https://i.imgur.com/9eDm6NQ.png" width="450" alt="Installing Sketch plugin" title="Installing Sketch plugin" />

Install `story2sketch`:

```sh
npm i story2sketch -g
```

Run `story2sketch`, pointing towards a Storybook iframe URL. You can find an existing iframe URL in Storybook by clicking 'Open canvas in new tab':

![Open canvas in new tab](https://i.imgur.com/Vo87WSM.png)

See [configuration](#configuration) for more options, or if you have a lot of stories.

```sh
story2sketch --url https://localhost:9001/iframe.html --output stories.asketch.json
```

Import the generated file into Sketch via `Plugins > From *Almost* Sketch to Sketch` in Sketch menu bar.

![Using sketch plugin](https://i.imgur.com/aA94aNN.png)

Success!

### Storybook 3.x

If you're using Storybook 3.3 or above (but not Storybook 4 or above), you'll want to [take full control of your Storybook webpack.config.js](https://storybook.js.org/configurations/custom-webpack-config/#full-control-mode) if you haven't already done so, adding:

```js
module.exports = (storybookBaseConfig, configType) => {
  const newConfig = {
    ...storybookBaseConfig
  };

  // Add this:
  // Export bundles as libraries so we can access them on page scope.
  newConfig.output.library = "[name]";

  return newConfig;
};
```

Manually export the `getStorybook` function in your `./config/storybook/config.js` file:

```js
import { getStorybook } from "@storybook/react";

...

export { getStorybook }
```

Run story2sketch:

```sh
story2sketch --url https://localhost:9001/iframe.html --output stories.asketch.json
```

## Why?

As stated by [`react-sketchapp`](https://github.com/airbnb/react-sketchapp), it's complicated to manage assets in a design system. Many teams building design systems or component libraries already produce Sketch files for distributing designs and use [Storybook](https://storybook.js.org) to prototype and present the developed components. It can become difficult to keep designs up to date with the latest components, with designers ever playing catchup. `story2sketch` generates a Sketch file from your components via Storybook, so your Sketch designs always stay up to date.

<a name="configuration"><a/>

## Configuration

You can configure `story2sketch` using [the API](#api) via the CLI, configuring your `package.json` or adding a `story2sketch.config.js` file.

### CLI

Simply call `story2sketch` with options from the [API](#api).

```sh
$ story2sketch --stories all --output dist/great-ui.asketch.json
```

### package.json

Add the following to your package.json:

```json
{
  "story2sketch": {
    "stories": "all",
    "output": "dist/great-ui.asketch.json"
  }
}
```

### story2sketch.config.js

Create a file called `story2sketch.config.js` on the root of your project:

```js
module.exports = {
  output: "dist/great-ui.asketch.json",
  stories: "all"
};
```

<a name="api"></a>

## API

| Parameter           | Explanation                                                                                                                                                                    | Input Type        | Default                                                                             |
| ------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ | ----------------- | ----------------------------------------------------------------------------------- |
| output              | Specifies the filename for the generated asketch.json file or a folder when outputBy === 'kind'.                                                                               | string            | `"dist/stories.asketch.json"`                                                       |
| input               | The location of Storybook's generated iframe.html. Use this over `url` if possible for performance.                                                                            | string            | `"dist/iframe.html"`                                                                |
| url                 | Storybook iframe URL. Will end in `iframe.html`. Prefer `input` for performance if possible.                                                                                   | string            | `"http://localhost:9001/iframe.html"`                                               |
| stories             | Stories to extract from Storybook. You should probably override the default.                                                                                                   | object/string     | `"all"`                                                                             |
| concurrency         | Number of headless Chrome tabs to run in parallel. Defaults to number of threads available on your machine.                                                                    | integer           | dynamic                                                                             |
| symbolGutter        | Gutter to place between symbols in Sketch.                                                                                                                                     | integer           | `100`                                                                               |
| viewports           | Viewport configuration. Will be arranged left-to-right by width. Try to avoid changing the key, as this is used to identify the symbol.                                        | object            | Mobile viewport (320px wide) and desktop viewport (1920px wide). See example below. |
| querySelector       | Query selector to select your node on each page. Uses `document.querySelectorAll`.                                                                                             | string            | `"#root"`                                                                           |
| verbose             | Verbose logging output.                                                                                                                                                        | boolean           | `false`                                                                             |
| fixPseudo           | Attempt to insert real elements in place of pseudo-elements                                                                                                                    | boolean           | `false`                                                                             |
| puppeteerOptions    | Options to be passed directly to `puppeteer.launch`. See [puppeteer docs](https://github.com/GoogleChrome/puppeteer/blob/master/docs/api.md#puppeteerlaunchoptions) for usage. | object            | `{}`                                                                                |
| removePreviewMargin | Remove preview margin from the iframe body.                                                                                                                                    | boolean           | `true`                                                                              |
| layoutBy            | Group symbols in the sketch output by the "kind" or "group" key                                                                                                                | "kind" \| "group" | null                                                                                |
| outputBy            | Write multiple sketch files by "kind" or the "group" key                                                                                                                       | "kind" \| "group" | null                                                                                |

## Example configurations

### Basic

Automatically detect the stories, outputting two viewports for each story in a single Sketch file as symbols.

```js
module.exports = {
  output: "dist/great-ui.asketch.json",
  input: "dist/iframe.html", // Same as default
  pageTitle: "great-ui"
};
```

### Manual stories

Manually define stories to have granular control over what stories are output. This might help if you're getting empty output, since some stories may break story2sketch.

```js
module.exports = {
  stories: [
    {
      kind: "Buttons/Button",
      stories: [
        {
          name: "Button"
        }
      ]
    },
    {
      kind: "Buttons/ButtonGroup",
      stories: [
        {
          name: "Default",
          displayName: "Horizontal"
        },
        {
          name: "Vertical"
        }
      ]
    },
    {
      kind: "Table",
      stories: [
        {
          name: "Table"
        }
      ]
    }
  ]
};
```

### Custom viewports

Output symbols based on custom viewports:

```js
module.exports = {
  viewports: {
    narrow: {
      width: 320,
      height: 1200,
      symbolPrefix: "Mobile/"
    },
    standard: {
      width: 1920,
      height: 1200,
      symbolPrefix: "Desktop/"
    }
  }
};
```

### Split output into multiple files based on kind

Outputs one file for each Storybook "kind". Useful if managing large component libraries, allowing you to distribute smaller files.

```js
module.exports = {
  output: "dist", // Define output directory. File names are defined by "kind"
  outputBy: "kind" // Also supports "group", see below.
};
```

### Layout based on kind

Renders the sketch layout by kind, but keeps them in one file.

```js
module.exports = {
  layoutBy: "kind" // Also supports "group", see below.
};
```

### Split output into multiple files based on custom group

This example outputs two files based on a custom grouping: `dist/Buttons.asketch.json` and `dist/Data.asketch.json`.

```js
module.exports = {
  output: "dist",
  outputBy: "group",
  stories: [
    {
      group: "Buttons",
      kind: "Buttons/Button",
      stories: [
        {
          name: "Button"
        }
      ]
    },
    {
      group: "Buttons",
      kind: "Buttons/ButtonGroup",
      stories: [
        {
          name: "Default",
          displayName: "Horizontal"
        },
        {
          name: "Vertical"
        }
      ]
    },
    {
      group: "Data",
      kind: "Table",
      stories: [
        {
          name: "Table"
        }
      ]
    }
  ]
};
```

## Continuous Integration

If you want `story2sketch` to run in a CI environment you might have to add the following configuration to puppeteer in your `story2sketch.config.js`.

```js
module.exports = {
  puppeteerOptions: {
    args: ['--no-sandbox', '--disable-setuid-sandbox']
  },
  ...
};
```

<a name="questions"><a/>

## Questions

### Why does my stuff look bad?

If your stuff looks bad, either it's not supported by [`html-sketchapp`](https://github.com/html-sketchapp/html-sketchapp) yet (see [support here](https://github.com/html-sketchapp/html-sketchapp/wiki/What's-supported%3F)), or you need to [configure story2sketch](#configuration).

### Why don't you use `react-sketchapp` instead of `html-sketchapp`?

[`react-sketchapp`](https://github.com/airbnb/react-sketchapp) only supports React Native, or forces you to use React Native component naming conventions. [`html-sketchapp`](https://github.com/html-sketchapp/html-sketchapp) supports good ol' fashioned HTML, and doesn't care what web framework you're using.

### Can I use this on anything other than Storybook?

Not yet, but we have plans to add support for multiple and custom adaptors.


CHANGELOG.md
<a name="1.7.1"></a>
## [1.7.1](https://github.com/chrisvxd/story2sketch/compare/v1.7.0...v1.7.1) (2020-11-16)


### Bug Fixes

* address skipped layers by updating html-sketchapp to 4.4.1 ([8ab4dc7](https://github.com/chrisvxd/story2sketch/commit/8ab4dc7))



<a name="1.7.0"></a>
# [1.7.0](https://github.com/chrisvxd/story2sketch/compare/v1.6.1...v1.7.0) (2019-09-17)


### Bug Fixes

* fix a bug where storybook without any stories runs forever ([1ff355e](https://github.com/chrisvxd/story2sketch/commit/1ff355e))


### Features

* upgrade to html-sketchapp@4.3.0 ([a453cda](https://github.com/chrisvxd/story2sketch/commit/a453cda))
* use dynamic thread count for concurrency ([fbe3d79](https://github.com/chrisvxd/story2sketch/commit/fbe3d79))



<a name="1.6.1"></a>
## [1.6.1](https://github.com/chrisvxd/story2sketch/compare/v1.6.0...v1.6.1) (2019-07-12)


### Bug Fixes

* handle replacing multiple spaces and slashes within a story name ([9b58e3d](https://github.com/chrisvxd/story2sketch/commit/9b58e3d))



<a name="1.6.0"></a>
# [1.6.0](https://github.com/chrisvxd/story2sketch/compare/v1.5.0...v1.6.0) (2019-05-20)


### Features

* upgrade to html-sketchapp@4.2.0 ([4db4157](https://github.com/chrisvxd/story2sketch/commit/4db4157))



<a name="1.5.0"></a>
# [1.5.0](https://github.com/chrisvxd/story2sketch/compare/v1.4.0...v1.5.0) (2019-04-22)


### Bug Fixes

* correct logs when outputting by group ([bb67cd1](https://github.com/chrisvxd/story2sketch/commit/bb67cd1))
* ensure outputBy param works correctly ([11fd7f4](https://github.com/chrisvxd/story2sketch/commit/11fd7f4))


### Features

* add support for grouping by symbols by "group" key ([f2a9ee0](https://github.com/chrisvxd/story2sketch/commit/f2a9ee0))
* add support for outputting multiple files by kind, or laying out stories by kind ([6290cae](https://github.com/chrisvxd/story2sketch/commit/6290cae))



<a name="1.4.0"></a>
# [1.4.0](https://github.com/chrisvxd/story2sketch/compare/v1.3.1...v1.4.0) (2018-12-03)


### Features

* add opt-in fix for pseudo-elements ([b1eb6a5](https://github.com/chrisvxd/story2sketch/commit/b1eb6a5))
* remove iframe body margin by default ([5a4b1e9](https://github.com/chrisvxd/story2sketch/commit/5a4b1e9))



<a name="1.3.1"></a>
## [1.3.1](https://github.com/chrisvxd/story2sketch/compare/v1.3.0...v1.3.1) (2018-10-15)


### Bug Fixes

* ensure preview is object with getStorybook() for storybook@3 ([1fe6409](https://github.com/chrisvxd/story2sketch/commit/1fe6409))



<a name="1.3.0"></a>
# [1.3.0](https://github.com/chrisvxd/story2sketch/compare/v1.2.0...v1.3.0) (2018-08-08)


### Features

* update to html-sketchapp@4.0.0 ([cfcf827](https://github.com/chrisvxd/story2sketch/commit/cfcf827))



<a name="1.2.0"></a>
# [1.2.0](https://github.com/chrisvxd/story2sketch/compare/v1.1.0...v1.2.0) (2018-08-08)


### Features

* add forward compatibility for storybook@4 ([fc39738](https://github.com/chrisvxd/story2sketch/commit/fc39738))
* add support for puppeteer.launch options ([66a8e57](https://github.com/chrisvxd/story2sketch/commit/66a8e57))
* upgrade to html-sketchapp@3.3.1 ([51ef691](https://github.com/chrisvxd/story2sketch/commit/51ef691))



<a name="1.1.0"></a>
# [1.1.0](https://github.com/chrisvxd/story2sketch/compare/v1.0.1...v1.1.0) (2018-06-19)


### Features

* add better error handling if story doesn't exist ([54bdfc1](https://github.com/chrisvxd/story2sketch/commit/54bdfc1))
* upgrade to html-sketchapp@3.2.0 ([61be07b](https://github.com/chrisvxd/story2sketch/commit/61be07b))



<a name="1.0.1"></a>
## [1.0.1](https://github.com/chrisvxd/story2sketch/compare/v1.0.0...v1.0.1) (2018-05-09)


### Bug Fixes

* order symbols correctly when using concurrency ([758f0da](https://github.com/chrisvxd/story2sketch/commit/758f0da))



<a name="1.0.0"></a>
# [1.0.0](https://github.com/chrisvxd/story2sketch/compare/v0.3.0...v1.0.0) (2018-05-01)


### Bug Fixes

* add missing babel-runtime dependency ([6f9155f](https://github.com/chrisvxd/story2sketch/commit/6f9155f))
* use and document correct querySelector default ([c879c0b](https://github.com/chrisvxd/story2sketch/commit/c879c0b))


### Features

* add "viewport" config for totally custom viewports ([6d42d50](https://github.com/chrisvxd/story2sketch/commit/6d42d50))
* upgrade to html-sketchapp@3.1.0 ([637aab9](https://github.com/chrisvxd/story2sketch/commit/637aab9))


### BREAKING CHANGES

* remove `narrowViewport` and `standardViewport` configuration options. Use
`viewports` instead.



<a name="0.3.0"></a>
# [0.3.0](https://github.com/chrisvxd/story2sketch/compare/v0.2.1...v0.3.0) (2018-04-24)


### Features

* generate layer name based on id, class or node type ([008dd16](https://github.com/chrisvxd/story2sketch/commit/008dd16))
* group Sketch layers based on DOM tree ([0c3947d](https://github.com/chrisvxd/story2sketch/commit/0c3947d))
* upgrade to html-sketchapp@3.0.2 ([07375cb](https://github.com/chrisvxd/story2sketch/commit/07375cb))



<a name="0.2.1"></a>
## [0.2.1](https://github.com/chrisvxd/story2sketch/compare/v0.2.0...v0.2.1) (2018-03-26)


### Bug Fixes

* ensure standard viewport output uses correct emoji ([fb4ab18](https://github.com/chrisvxd/story2sketch/commit/fb4ab18))



<a name="0.2.0"></a>
# [0.2.0](https://github.com/chrisvxd/story2sketch/compare/0.1.4...v0.2.0) (2018-03-26)


### Features

* upgrade html-sketchapp to 2.0 ([23001fe](https://github.com/chrisvxd/story2sketch/commit/23001fe))



<a name="0.1.4"></a>
## [0.1.4](https://github.com/chrisvxd/story2sketch/compare/0.1.3...0.1.4) (2018-01-25)


### Bug Fixes

* ensure prod dependencies are installed when using npm ([ad27893](https://github.com/chrisvxd/story2sketch/commit/ad27893))



<a name="0.1.3"></a>
## [0.1.3](https://github.com/chrisvxd/story2sketch/compare/0.1.2...0.1.3) (2018-01-25)


### Bug Fixes

* ensure lib is correctly published ([93e7b7a](https://github.com/chrisvxd/story2sketch/commit/93e7b7a))



<a name="0.1.2"></a>

src/server/Story2sketch.js
/* eslint-disable no-console*/

import puppeteer from "puppeteer";
import fs from "fs";
import path from "path";
import mkdirp from "mkdirp";
import chalk from "chalk";
import ProgressBar from "progress";

import getStorybook from "./getStorybook";
import PagePool from "./PagePool.js";

const defaultConcurrency = 4;
const defaultSymbolGutter = 100;

export default class Story2sketch {
  constructor({
    output = `${process.cwd()}/dist/stories.asketch.json`,
    url = "http://localhost:9001/iframe.html",
    concurrency = defaultConcurrency,
    pageTitle = "Stories",
    viewports = {
      narrow: {
        width: 320,
        height: 1200,
        symbolPrefix: "📱 "
      },
      standard: {
        width: 1920,
        height: 1200,
        symbolPrefix: "🖥 "
      }
    },
    symbolGutter = defaultSymbolGutter,
    querySelector = "#root",
    verbose = false,
    fixPseudo = false,
    stories,
    puppeteerOptions = {},
    removePreviewMargin = true,
    layoutBy = null,
    outputBy = null
  }) {
    this.output = output;
    this.url = url;
    this.concurrency = concurrency;
    this.pageTitle = pageTitle;
    this.symbolGutter = symbolGutter;
    this.viewports = viewports;
    this.querySelector = querySelector;
    this.stories = stories;
    this.verbose = verbose;
    this.fixPseudo = fixPseudo;
    this.layoutBy = layoutBy;
    this.outputBy = outputBy;
    this.removePreviewMargin = removePreviewMargin === true;
    this.puppeteerOptions = puppeteerOptions;

    // Sort viewports by width
    this.sortedViewports = Object.keys(viewports)
      .map(key => ({
        ...viewports[key],
        id: key
      }))
      .sort((a, b) => a.width > b.width);
  }

  reset() {
    this.symbolsByViewport = {};
    this.symbolsByGrouping = {};
    this.widestByViewport = {};
    this.tallestByStory = {};
    this.processedStories = 0;
    this.storyCount = 0;
    this.sketchPage = {};
  }

  async init() {
    this.browser = await puppeteer.launch(this.puppeteerOptions);

    if (!this.stories || this.stories === "all") {
      if (this.verbose) {
        console.log(chalk.gray("Detecting stories..."));
      }

      this.stories = await getStorybook(this.browser, this.url);
    }

    this.reset();

    this.pagePool = await this.createPagePool();

    this.progressBar = new ProgressBar(
      "[:bar] :percent (:current/:total) :etas remaining",
      {
        total: this.storyCount,
        story: "",
        complete: "#"
      }
    );

    this.sketchPage = await this.getSketchPage();

    console.log(`Processing ${this.storyCount} stories...`);
  }

  // NB The only reason this needs to run in chrome is because html-sketchapp
  // uses imports/exports and therefore won't compile for node. html-sketchapp
  // either needs to compile down, or we can webpack the server bundle.
  async getSketchPage() {
    if (this.verbose) {
      console.log(chalk.gray("Getting sketch page..."));
    }

    const page = await this.browser.newPage();

    await page.goto(this.url, {
      waitUntil: "networkidle2"
    });

    await page.addScriptTag({
      path: `${__dirname}/../browser/page2layers.bundle.js`
    });

    const params = JSON.stringify({
      title: this.pageTitle,
      width: 1920,
      height: 5000
    });

    const sketchPage = await page.evaluate(`
      page2layers
      .getPage(${params})
    `);

    return sketchPage;
  }

  getGrouping({ group, kind }) {
    let grouping = "all";

    if (this.outputBy === "kind") {
      grouping = kind;
    } else if (this.outputBy === "group") {
      grouping = group;
    }

    if (this.layoutBy === "kind") {
      grouping = kind;
    } else if (this.layoutBy === "group") {
      grouping = group;
    }

    return grouping;
  }

  async createPagePool() {
    const pagePool = new PagePool(this.browser, this.concurrency);

    await pagePool.init();

    for (const { group, kind, stories } of this.stories) {
      const grouping = this.getGrouping({ group, kind });

      this.symbolsByGrouping[grouping] = {};

      for (const { id } of this.sortedViewports) {
        this.symbolsByGrouping[grouping][id] = Array(stories.length);
      }

      for (const story of stories) {
        const storyIndex = this.storyCount;

        this.storyCount++;

        pagePool.queue(async page => {
          const symbolByViewport = await this.getSymbolsForStory({
            page,
            kind,
            story
          });

          let tallest = 0;

          for (const viewportKey in symbolByViewport) {
            const symbol = symbolByViewport[viewportKey];

            tallest = Math.max(tallest, symbol.frame.height);

            this.widestByViewport[viewportKey] = Math.max(
              this.widestByViewport[viewportKey] || 0,
              symbol.frame.width
            );

            // Assign by index to retain the order of the symbols
            this.symbolsByViewport[viewportKey][storyIndex] = symbol;
            this.symbolsByGrouping[grouping][viewportKey][storyIndex] = symbol;
          }

          this.tallestByStory[storyIndex] = tallest;
        });
      }
    }

    // Initialize an array per viewport based on the number of stories
    for (const { id } of this.sortedViewports) {
      this.symbolsByViewport[id] = Array(this.storyCount);
    }

    return pagePool;
  }

  async getSymbolsForStory({ page, kind, story }) {
    const builtUrl = `${this.url}?selectedKind=${encodeURIComponent(
      kind
    )}&selectedStory=${encodeURIComponent(story.name)}`;

    await page.goto(builtUrl, {
      waitUntil: "networkidle2"
    });

    await page.addScriptTag({
      path: `${__dirname}/../browser/page2layers.bundle.js`
    });

    const name = `${kind}/${story.displayName || story.name}`;
    const symbolByViewport = {};

    for (const { symbolPrefix = "", id, width, height } of this
      .sortedViewports) {
      await page.setViewport({ width, height });

      // Only prefix if symbolPrefix is defined
      const params = JSON.stringify({
        name: `${symbolPrefix}${name}`,
        querySelector: this.querySelector,
        fixPseudo: this.fixPseudo,
        removePreviewMargin: this.removePreviewMargin
      });

      // JSON.parse + JSON.stringify hack was originally used until
      // https://github.com/GoogleChrome/puppeteer/issues/1510 was fixed, but
      // it still results in better performance.
      const symbolJson = await page.evaluate(`
        JSON.stringify(
          page2layers
          .getSymbol(${params})
        );
        `);

      const symbol = JSON.parse(symbolJson);

      if (symbol) {
        symbol.symbolID = `${name}:${id}`;

        symbolByViewport[id] = symbol;
      } else {
        console.warn(
          chalk.yellow(
            ` WARNING: No matching node for "${name}" using querySelector ${
              this.querySelector
            } on viewport ${id}`
          )
        );
      }
    }

    this.logProgress(name);

    return symbolByViewport;
  }

  logProgress(name) {
    this.processedStories += 1;

    if (this.verbose) {
      console.log(
        `${chalk.green("✓")} ${chalk.gray(`Exported`)} ${chalk.bold(`${name}`)}`
      );
    } else {
      this.progressBar.tick({
        processed: this.processedStories,
        story: chalk.bold(name)
      });
    }
  }

  positionSymbols(symbolsByViewport, xOffsetStart = 0) {
    let layers = [];
    let xOffset = xOffsetStart;

    for (const { id } of this.sortedViewports) {
      let yOffset = 0;

      const symbols = symbolsByViewport[id];

      for (const [index, symbol] of symbols.entries()) {
        // Skip failed symbols
        if (symbol) {
          symbol.frame.x = xOffset;
          symbol.frame.y = yOffset;
          layers.push(symbol);
          yOffset += this.tallestByStory[index] + this.symbolGutter;
        }
      }

      xOffset += this.widestByViewport[id] + this.symbolGutter;
    }

    return layers;
  }

  positionSymbolsByViewport() {
    this.sketchPage.layers = [
      ...this.sketchPage.layers,
      ...this.positionSymbols(this.symbolsByViewport)
    ];
  }

  positionSymbolsByGrouping() {
    const width = this.sortedViewports.reduce(
      (totalWidth, viewport) =>
        totalWidth + this.widestByViewport[viewport.id] + this.symbolGutter,
      0
    );

    for (const [index, grouping] of Object.keys(
      this.symbolsByGrouping
    ).entries()) {
      const symbolsByViewport = this.symbolsByGrouping[grouping];

      this.sketchPage.layers = [
        ...this.sketchPage.layers,
        ...this.positionSymbols(symbolsByViewport, width * index)
      ];
    }
  }

  writeByGrouping() {
    mkdirp(this.output);

    for (const grouping of Object.keys(this.symbolsByGrouping)) {
      const sketchPage = {
        ...this.sketchPage,
        layers: this.positionSymbols(this.symbolsByGrouping[grouping])
      };
      const filename = `${grouping
        .replace(/ /g, "_")
        .replace(/\//g, "+")}.asketch.json`;

      const outputPath = path.join(this.output, filename);

      fs.writeFileSync(outputPath, JSON.stringify(sketchPage));

      console.log(
        chalk.green(
          `Wrote stories by ${this.outputBy} to ${chalk.white.bold(outputPath)}`
        )
      );
    }
  }

  async execute() {
    process.on("SIGINT", () => {
      this.browser.close();

      this.exited = true;

      process.exit();
    });

    await this.pagePool.execute().catch(error => {
      // Suppress errors if user exited process.
      if (!this.exited) {
        console.error(error);
      }
    });

    if (this.outputBy) {
      this.writeByGrouping();
    } else {
      if (this.layoutBy) {
        this.positionSymbolsByGrouping();
      } else {
        this.positionSymbolsByViewport();
      }

      fs.writeFileSync(this.output, JSON.stringify(this.sketchPage));
    }

    console.log(
      chalk.green(
        `Success! ${
          this.processedStories
        } stories written to ${chalk.white.bold(this.output)}`
      )
    );

    this.browser.close();
    process.exit();
  }
}
